### PR TITLE
Add interpolation option for ObserveFields

### PR DESCRIPTION
**Pull request by nikwit**
## Proposed changes

an optional Mesh can now be passed to the ObserveFields constructor to which all specified tensors are interpolated

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [ ] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments
I found Test_ObserveFields.cpp to be almost completely unreadable and quite rigid so I was unable to add the interpolation in a more elegant way. Maybe you can think of a nicer way without completely rewriting the unit test.

The test now basically recreates the internals of the interpolation of ObserveFields and checks it against the final result of ObserveFields. Maybe it would make more sense for the unit test to check against some analytical solution.

### CODE REVIEW

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: typo: "the" or "all"

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: typo: is run -> runs

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp:142]: I think we should make sure the user understand that if they specify this option, the _original_ data will be lost and only the interpolated data is written to disk. We can do these things:

- Add a `static type suggested_value() noexcept { return {}; }` so the executable prints a note on startup that this option differs from the suggestion. I think this would be sufficient warning to the user.
- Add a note to this help string that the user can add the event multiple times, with and without the interpolating mesh, if they want both the original and the interpolated data written out, possibly at different triggers.

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: This shouldn't compile.. does it? It should be `suggested_value`

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp:167]: not sure why this is taken by const-ref and copied. Could you change it to taken-by-value and `std::move` it instead? You can also `std::move` the mesh. (If clang-tidy says the move is trivial, just silence it)

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: delete `not`

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: [optional] I prefer the more expressive `interpolation_mesh_.has_value()`

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: Prefer `++i` over `i++`. It's good to get in the habit of prefix-incrementing to avoid copies in cases where they actually matter.

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: `const auto&` (the collocation points are statically-cached)

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: I suggest to wrap the body of the dimension-loop in another condition that checks `slices.at(i) != interpolation_slices.at(i)`

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: This will copy the `error` even if no interpolation is performed. Same for the `tensor[I]` in the line 245. Not sure if we care though..

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp:27]: Also include `<optional>` and `Parallel/PupStlCpp17.hpp`

File: tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp
Comment (nilsvu) [inline tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp]: You can grab `System::volume_dim` here, no need to add the template parameter

File: tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp
Comment (nilsvu) [inline tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp]: Here, too, you can grab the `System::volume_dim`

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nikwit) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp:215]: Would you multiply `components.reserve` by `interpolation_mesh->extents() / mesh.extents()` to account for the different size required for the interpolating mesh?

Comment (nilsvu) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp:215]: The `reserve` only deals with the number of components, not the number of grid points per component

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nikwit) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp]: I see it as plausible that the user might specify the default grid as the interpolation grid, so we could add `if (interpolation_mesh_.has_value() && interpolation_mesh.value() != mesh)`.

Otherwise you are addressing the case that the user would request an interpolation to a mesh that is different from the default mesh in certain dimensions only. Is this ever going to happen and worth including the condition check?

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nikwit) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp:215]: ah of course

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp]: I think it's perfectly possible the mesh only differs in a particular dimension from the simulation mesh. For example if basis and quadrature are the same, we're upsampling to a constant num points and the domain is refined such that some elements have exactly that num points in some dimension.

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nikwit) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp:167]: I can do that. This will also make the check for duplicates redundant (and impossible as we are moving from std::vector into std::unordered_set so duplicate information is lost). I will  remove the corresponding unit test failure case.

Comment (nilsvu) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp:167]: Oh I didn't realize this is stored as an `unordered_set` but the option parses as a `vector`. Probably best to leave as-is then (unless we added option-parsing for `unordered_set` at some point)

Comment (nilsvu) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp:167]: I suggest you don't bother with this for this PR

Comment (nikwit) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp:167]: ha, too late. the implementation I have now should still be option parsable though as the argument is still a stringvector

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (wthrowe) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: Remove "default".

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (wthrowe) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp:142]: Do you really think that if people request their data to be interpolated that they will be surprised when they don't get data they didn't request?  The documentation is good, but I feel that in general we are using suggestions too much.

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (wthrowe) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: Don't change this.  You've removed error checking.

File: tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp
Comment (wthrowe) [inline tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp]: This section of the test appears to be copy-pasted from the code it is testing.  Can you do an independent test?  Maybe construct linear data and check that is is interpolated correctly.

File: tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp
Comment (wthrowe) [inline tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp]: Test nonuniform meshes.  We can't parse them yet, but we don't want to have to change this class when we can.

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (wthrowe) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: "The recorded quantities are" (or similar).

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsdeppe) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp:142]: I'm with @wthrowe. Observing data is opt-in and if you choose to write data to a different grid then nobody should be surprised that they got data on the grid they requested. If someone is surprised then I think it's a lesson they should learn, even if it's learned the hard way once. I really don't need the code yelling at me all the time "Hey, you aren't observing on the same grid you're solving the PDEs on!" It's not like I get incorrect answers if I observe on a different grid. `suggested_values` should only be used in cases where if you don't use the suggested value the simulation is likely to fail or behave incorrectly. Anything else, make people choose what they want. Furthermore, if you interpolate to a mesh with the same or higher-degree basis you haven't actually lost anything (unless you interpolate across a discontinuity but that's a different problem completely).

Adding a note about having multiple events: I don't think we need to start documenting options of events for how events work. Having documentation on the class/event itself that it can be added more than once is okay, but I think it would make more sense to have a tutorial on events. Additionally, at least in the evolution code the observation example input file does several observations at once.

Comment (nilsvu) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp:142]: Ok works for me

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: Keep the default-construction please. Everything has to be default-constructible for Charm++. I guess removing the braces still default-constructs this but I think we prefer being explicitly about it (and clang-tidy does, too)

**Review by nilsvu — COMMENTED** [review]
I'm happy with squashing this once @wthrowe's comments are addressed (no need to wait for my re-approval)

**Comment by nikwit** [comment]
@wthrowe , after a call with @nilsleiffischer we decided that there is little point in checking the currently existing unit test by manually interpolating data. This would mostly boil down to testing the `interpolation_matrix` function which isn't the purpose of this PR or the unit test. 

Instead, we thought it would be best to isolate this logic of interpolating from one mesh to another in a separate function which I have added in the last commit. This function would also be used by the executable I'm currently working on which does the interpolation for existing data in post-processing. Testing this isolated function is far easier as it does not need to use the action testing framework and does not rely on the rather rigid, existing unit test. 

If this is fine with you, I will add an appropriate unit test for this function.

File: src/NumericalAlgorithms/Spectral/Spectral.cpp
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Spectral/Spectral.cpp]: How is this different from `RegularGrid` in `NumericalAlgorithms/Interpolation/RegularGridInterpolant.hpp`?

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: This was deleted in #2661, you can just drop this change

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: ..except for a copy. Probably irrelevant, since the non-solution tensors have to be copied anyway and optimizing this just for the errors seems premature. Please add a note to the comment though.

File: tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp
Comment (nilsvu) [inline tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp]: `static constexpr bool volume_dim`

File: tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp
Comment (nilsvu) [inline tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp]: I think just call `test_observe` directly here instead of adding the case for the empty string.

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: I think this note is helpful in this context, and if I understood @nilsdeppe correctly he didn't object, just suggested we write more documentation about events in a dev guide. So please drop this fixup commit.

**Review by nilsvu — CHANGES_REQUESTED** [review]
Could you start the commit messages with a capital letter? (please excuse my nitpicking 😄)

File: tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp
Comment (nilsvu) [inline tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp]: You can wrap each of these tests in a scope:

```
{
  INFO("Interpolate to coarser grid");
  // ...
}
```

Then you can use the same variable names everywhere (don't need to add `fewer_points_` prefixes and such) and get the added benefit that when a test fails the `INFO` gets printed.

**Review by nilsvu — CHANGES_REQUESTED** [review]
I'm happy with squashing everything. I suggest you make the changes I suggested and squash everything now, seems easier than reviewing fixups that change things back and forth :)

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: Delete this line

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsvu) [inline src/ParallelAlgorithms/Events/ObserveFields.hpp]: Delete this line

File: tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp
Comment (nilsvu) [inline tests/Unit/ParallelAlgorithms/Events/Test_ObserveFields.cpp]: Delete extra _

**Review by nilsvu — CHANGES_REQUESTED** [review]
LGTM, please squash these couple of things in directly.

You have a failure on CI, looks like you have to link in the library for the interpolant. List it under `INTERFACE` in `target_link_library` because only a header needs it.

File: src/NumericalAlgorithms/Interpolation/RegularGridInterpolant.hpp
Comment (nilsdeppe) [inline src/NumericalAlgorithms/Interpolation/RegularGridInterpolant.hpp:58]: you can remove the `makearray` include as well :)

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsdeppe) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp]: I think with a rebase now this default should disappear complete. I might have missed it but we should delete it here

**Review by nilsdeppe — COMMENTED** [review]
A couple minor suggestions. This looks good :) Nice work! No need to wait for me to approve for this to get merged :)

File: src/ParallelAlgorithms/Events/ObserveFields.hpp
Comment (nilsleiffischer) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp:163]: I don't understand why this compiles. Is the default value never used and thus the compiler never tries to evaluate this piece of code? In any case, please delete the default.

Comment (nilsvu) [reply src/ParallelAlgorithms/Events/ObserveFields.hpp]: I don't understand why this compiles. Is the default value never used and thus the compiler never tries to evaluate this piece of code? In any case, please delete the default.

**Review by nilsvu — CHANGES_REQUESTED** [review]
Please squash directly

**Review by wthrowe — APPROVED** [review]

**Review by nilsvu — APPROVED** [review]

**Pull request merged**
